src/util/importJSON.ts

```diff
@@ -167,52 +156,56 @@ const saveThoughts = (
           updatedBy,
         )
 
-        // TODO: remove mutations
-        contextIndexUpdates[contextEncoded] = parent
-        // `Parent` entry for the newly created thought.
-        contextIndexUpdates[newThought.id] = newThought
-
-        thoughtIndexUpdates[hashThought(nonDuplicateValue)] = lexeme
+        return {
+          thoughtIndexUpdates: {
+            [hashThought(nonDuplicateValue)]: lexeme,
+          },
+          contextIndexUpdates: {
+            [contextEncoded]: parent,
+            [newThought.id]: newThought,
+          },
+        }
       }
 
+      const insertUpdates = !skipLevel ? insert() : null
+
+      const updatedState = insertUpdates
+        ? {
+            ...stateNewBeforeInsert,
+            thoughts: {
+              ...stateNewBeforeInsert.thoughts,
+              contextIndex: mergeUpdates(stateNewBeforeInsert.thoughts.contextIndex, insertUpdates.contextIndexUpdates),
+              thoughtIndex: mergeUpdates(stateNewBeforeInsert.thoughts.thoughtIndex, insertUpdates.thoughtIndexUpdates),
+            },
+          }
+        : stateNewBeforeInsert
+
       const updatedDuplicateIndex = {
         ...accum.duplicateIndex,
         [hashedValue]: duplicateValueCount ? duplicateValueCount + 1 : 1,
       }
 
-      const updatedState: State = {
-        ...state,
-        thoughts: {
-          ...state.thoughts,
-          contextIndex: {
-            ...state.thoughts.contextIndex,
-            ...contextIndexUpdates,
-          },
-          thoughtIndex: {
-            ...state.thoughts.thoughtIndex,
-            ...thoughtIndexUpdates,
-          },
-        },
-      }
-
       /**
        *
        */
       const getLastAddedChild = () => {
-        const parent = contextIndexUpdates[contextEncoded]
+        const parent = updatedState.thoughts.contextIndex[contextEncoded]
         return (childIdsToThoughts(updatedState, parent.children) ?? []).find(
           child => child.value === nonDuplicateValue,
         )
       }
 
+      const updatedContextIndexUpdates = mergeUpdates(accum.contextIndex, insertUpdates?.contextIndexUpdates || {})
+      const udpatedThoughtIndexUpdates = mergeUpdates(accum.thoughtIndex, insertUpdates?.thoughtIndexUpdates || {})
```

Comment:
These names are misleading, because the values are proper `contextIndex` and `thoughtIndex` types, not updates which are typically nullable.

Reply:
Since adding new entries to the thought indices are kind of like updates too, so I went with that naming. However we can choose not to use `updates` as the suffix for these variable names.

Reply:
If the updates are merged into a full `contextIndex` (i.e. taken from `state.thoughts.contextIndex`), then we would consider the result to be a new `contextIndex`. If the updates are merged into another set of updates, then we would consider the result to be strictly updates. 

Note that `mergeUpdates` is *only* used to merge into a full `contextIndex`. Otherwise we would not need to delete falsey values and could simply use `{ ...contextIndexUpdates1, ...contextIndexUpdates2 }`.